src/lib/mongoBackend/mongoQueryContext.cpp

```diff
@@ -38,6 +38,268 @@
+      {
+        std::string attrName = cerV.get(ix)->contextElement.contextAttributeVector.get(jx)->name;
+        if (attrName == craP->name)
+        {
+          /* In this case, the attribute has been already found in local database. CPr is unnecessary */
+          return;
+        }
+      }
+      /* Reached this point, no attribute was found, so adding it with corresponding CPr info */
+      ContextAttribute* caP = new ContextAttribute(craP->name, "", "");
+      caP->providingApplication = pa;
+      cerV.get(ix)->contextElement.contextAttributeVector.push_back(caP);
+      return;
+    }
+  }
+
+  /* Reached this point, it means that the cerV doesn't contain a proper CER, so we create it */
+  ContextElementResponse* cerP = new ContextElementResponse();
+  cerP->contextElement.entityId.id = enP->id;
+  cerP->contextElement.entityId.type = enP->type;
+  cerP->contextElement.entityId.isPattern = "false";
+
+  cerP->statusCode.fill(SccOk);
+
+  ContextAttribute* caP = new ContextAttribute(craP->name, "", "");
+  caP->providingApplication = pa;
+  cerP->contextElement.contextAttributeVector.push_back(caP);
+
+  cerV.push_back(cerP);
+}
+
+
+/* ****************************************************************************
+*
+* matchEntityInCrr -
+*
+*/
+bool matchEntityInCrr(ContextRegistration& cr, EntityId* enP)
+{
+  for (unsigned int ix = 0; ix < cr.entityIdVector.size(); ++ix)
+  {
+    EntityId* crEnP = cr.entityIdVector.get(ix);
+    if (matchEntity(crEnP, enP))
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+
+/* ****************************************************************************
+*
+* addContextProviders -
+*
+*
+* This functions takes a CRR vector and adds the Context Providers in the CER vector
+* (except the ones corresponding to some locally found attribute, i.e. info already in the
+* CER vector)
+*
+* The enP paramter is optional. If not NULL, then before adding a CPr the function check that the
+* containting CRR match the entiy (this is used for funcionality related with "generic queries", see
+* processGenericEntities() function)
+*
+*/
+void addContextProviders(ContextElementResponseVector& cerV, ContextRegistrationResponseVector& crrV, EntityId* enP = NULL)
+{
+  for (unsigned int ix = 0; ix < crrV.size(); ++ix)
+  {
+    ContextRegistration cr = crrV.get(ix)->contextRegistration;
+
+    /* In the case a "filtering" entity was provided, check that the current CRR match or skip to need CRR */
+    if (enP != NULL && !matchEntityInCrr(cr, enP)) {
+      continue;
+    }
+
+    if (cr.contextRegistrationAttributeVector.size() == 0)
+    {
+      /* Registration without attributes */
+      for (unsigned int jx = 0; jx < cr.entityIdVector.size(); ++jx)
+      {
+        addContextProviderEntity(cerV, cr.entityIdVector.get(jx), cr.providingApplication.get());
+      }
+    }
+    else
+    {
+      /* Registration with attributes */
+      for (unsigned int jx = 0; jx < cr.entityIdVector.size(); ++jx)
```

Comment:
I try to use indexes (indices?) to illustrate a little what they are counting,
For entityIdVector I would use *eIx*, for contextRegistrationAttributeVector *craIx*.
It's a little nicer than jx, kx ...